forest/src/daemon.rs

```diff
@@ -339,6 +294,56 @@ pub(super) async fn start(config: Config) {
         None
     };
 
+    select! {
+        () = sync_from_snapshot(&config, &state_manager).fuse() => {},
+        _ = ctrlc_oneshot => {
+            return;
+        },
+    }
+
+    // Terminate if no snapshot is provided or DB isn't recent enough
+    match chain_store.heaviest_tipset().await {
+        None => {
+            cli_error_and_die(
+                "Forest cannot sync without a snapshot. Download a snapshot from a trusted source and import with --import-snapshot=[file]",
+                1,
+            );
+        }
+        Some(tipset) => {
+            let epoch = tipset.epoch();
+            let nv = config.chain.network_version(epoch);
+            if nv < NetworkVersion::V16 {
+                cli_error_and_die(
+                    "Database too old. Download a snapshot from a trusted source and import with --import-snapshot=[file]",
+                    1,
+                );
+            }
+        }
+    }
+
+    // Halt
+    if config.client.halt_after_import {
+        info!("Forest finish shutdown");
+        return;
+    }
+
+    // Fetch and ensure verification keys are downloaded
+    if cns::FETCH_PARAMS {
+        use forest_paramfetch::{
+            get_params_default, set_proofs_parameter_cache_dir_env, SectorSizeOpt,
+        };
+        set_proofs_parameter_cache_dir_env(&config.client.data_dir);
+
+        get_params_default(&config.client.data_dir, SectorSizeOpt::Keys, false)
+            .await
+            .unwrap();
+    }
+
+    // Start services
+    let p2p_task = task::spawn(async {
```

Comment:
An edge case, but IMO worth considering. What will happen if we call some of the commands like `forest net peers` while the snapshot is loading? The p2p service was initialized before RPC, right now I'm not entirely sure what will happen and do we have a nice error handling.

Reply:
Good question. Right `forest net peers` blocks until the p2p service is up and running. I feel this is alright but I haven't given it much thought.

Reply:
I'd say an immediate answer that the Forest service is busy may be more convenient for whoever is integrating with it, but it's fine for now, we should revisit those calls later on when covering them with a nice suite of tests.